### ISSUE TITLE
Media: images uploaded via an image block are unattached

### ISSUE THREAD

**Issue by sophiegyo** [Automattic/wp-calypso#44427]
TL;DR: When uploading an image via image block on Simple sites on dotcom, the upload is listed as "unattached" in the media library.
This seems to be exclusive to dotcom Simple sites. 

#### Steps to reproduce

1. Starting with a Simple dotcom site: https://wordpress.com/block-editor/post/SITEADDRESS
2. Insert an image block
3. Upload image from computer
4. Save draft or publish
5. Check media library
6. Newly uploaded image is "unattached".

#### What I expected
The image should be attached to the post it was uploaded to.

#### What happened instead
The image is not attached

#### Browser / OS version
Mac OS 10.14.6
Firefox 77.0.1

#### Screenshot / Video
Test 3: Gutenberg in both Calypso and wp-admin (my own test site): https://d.pr/v/Jsjiyj
Test 4: Simple site, classic editor https://d.pr/v/iIh7kp
Test 5: Selfhosted, Gutenberg https://d.pr/v/zB8Ka3
Test 6: AT, Gutenberg https://d.pr/v/R9aRlA

#### Context / Source
Chat/ticket. Ticket follow-up 3169671-zen
 #user-report

Checking over search results in MGS, it looks like there have been a few instances recently of people deleting their unattached images, then realising those images were in fact included in posts or pages, so I'd call this a #painpoint
Related issue: #26908

**Comment by arunsathiya** [comment]
User in original report would like to followup when it's resolved.

**Comment by sirreal** [comment]
This is a great report, thank you. The videos are really helpful to understand the issue. I've added this to the Create maintenance board.

**Comment by marutim** [comment]
The user checked about the status of the issue in the chat (24979468-hc) today

**Comment by synora** [comment]
Removed stale label as I'm still able to replicate this issue.

**Comment by supernovia** [comment]
This gets stranger. I was trying to figure out a workaround to help an [FSE hopeful](https://wordpress.com/forums/topic/experimenting-with-fse/) get their images attached. I published a new post with a gallery, then headed to the media library. If I filtered the grid view for "unattached" those new images don't show up, even though some known unattached images did. Bug must have been solved, right?

So I was going to have the user head to the list view and start reattaching images. I checked to be sure the attachment instructions were the same. My new images appeared there, even with the "unattached" filter active. And even if I tried to filter it like this: `/wp-admin/upload.php?mode=list&attachment-filter=detached`

It turns out my new images _were in fact not attached_. They just didn't show up in the "unattached" grid view. 🤔 

So then I tried attaching the images. I hovered, clicked Attach, and tried to pick my page. It all looked like it worked. But then the images were still showing up as unattached, even though the image is in fact listed as attached now when I view it individually. 🙃

**Comment by CraigG58** [comment]
This is a serious issue.

Until fixed, “Home”>”Media” to check status of file attachments = Completely useless, and more importantly dangerous in that those in use may be inadvertently and permanently deleted.

Any chance of bumping up the priority here? It's been 20 months since it was first reported, and no one is even assigned to resolve it yet.

**Comment by supernovia** [comment]
Noting discussion here: pc4f5j-1yn-p2

**Comment by supernovia** [comment]
Noting as well: while a standard WP install does seem to attach images, the unattached filter isn't working consistently. If I attach an image there, it still appears in an unattached list.

**Comment by formosattic** [comment]
>  This seems to be exclusive to dotcom Simple sites.

This still appears to be the case based on my testing.

I tried on self-hosted and atomic, with and without Gutenberg, and images were attached.

Noting that checking for unattached images is an advice shared in our doc, https://wordpress.com/support/media/view-all/#unattached-files , and we might indeed be sharing it more.

Reattaching media ([see guide](https://wordpress.com/support/media/view-all/#reattach-files)) can help but doesn't solve the root issue.

User report: 4915461-zd-woothemes

**Comment by CraigG58** [comment]
Just wanted to interject something...

When and if this bug is finally fixed, expecting users to re-attach media manually would be insane. I’ve no idea what an average might be, but my media library currently contains 781 files. That said, even if I was willing, it would take until Christmas (of 2025) to go through and correct everything myself.

_Surely this could be accomplished programmatically_ by inspecting existing pages and posts, then cross-referencing and updating media library attachment status accordingly. It could be an optional thing left up to the user to run, but in my opinion this definitely needs to be part of fixing the bug.

Thanks for your consideration.

**Comment by p-jackson** [comment]
Work-in-progress that fixes the underlying issue: D78584-code

Re-attaching existing media to posts will need to be done as a follow-up step. We can't know for sure that every image that's currently unattached is that way just because of this bug. It could also be because the file was originally uploaded via the media library (those images aren't attached to any post). So I don't think it's something we can do automatically for all sites. But it could be a one-off action we run on a per-site basis if the user asks.

Another option might be to provide a better tool to help users prune their media. The "Uploaded to" column is a good approximation of whether an image is currently in use, but it's not foolproof. An image could be used in multiple posts but still listed as "unattached". Maybe we could provide a filter that lets users see which images aren't currently used in pages/posts regardless of their attached status?

**Comment by CraigG58** [comment]
At the risk of repeating myself, please bear with me... I’m obviously not familiar with the internal workings of WordPress code, but I still can’t imagine why media file attachment status couldn’t be fixed programmatically after this bug is corrected, regardless of how it was uploaded (inside or outside the editor).

> We can't know for sure that every image that's currently unattached is that way just because of this bug.

That's true in part, but for a given WordPress account I would assume there is a Gutenberg (or Classic editor) source file on your servers for each and every every post/page, or perhaps they live in one large XML file. Either way, it seems this is all a “fix attachments” program of some sort would need to do:

1) Scan those source files to see if they reference any media files with extensions supported by WordPress (jpg, png, mp4, etc).

2) For each media file found referenced within the post/page source files, check its status in the library (attached/unattached).

3) If flagged as “unattached”, update the library to correct it as “attached”, along with the post/page name.

Seems pretty simple, or am I missing something?

Checking and updating attachment status of each media file is completely unrealistic to expect of users, since it would consume a lifetime in the case of large libraries like mine. It currently contains 772 media items with 418 flagged as unattached, and I’m positive nearly all of them are in use since I use great care in maintaining my library so as not to waste space.

Even if a user went forward with performing manual attachment status corrections, there's nothing to reference as the library listing is completely inaccurate at this time and therefore totally useless. That said, the only way a user could check the actual status of library files would be taking the same three steps described above, _manually in every post/page on their blog_. All due respect, but can you just imagine how long that would take?

Thanks once again for your consideration and support, I appreciate it!

**Comment by p-jackson** [comment]
Just deployed D78584-code which fixes the first issue. Newly uploaded images will now be attached to the correct post.

**Comment by CraigG58** [comment]
FYI, Just tried it by editing an existing post and it seems to work fine now. Thank you so much, I’ll keep an eye on this while writing future Posts and Pages!

However, now that this issue has been fixed, it would be the perfect time to correct the larger problem of existing media files being incorrectly flagged as “unattached”. I had suggested one possibility to take care of this in my last post, and was wondering if you’ve considered something along those lines?

Thanks again for squashing a very serious bug, but the media list in the Administrator dialog remains worthless for determining which files are indeed in use.

**Comment by p-jackson** [comment]
I've created a script that reattaches unattached images to the first post that references them: either in the post body or as the featured image. As [I mentioned above](https://github.com/Automattic/wp-calypso/issues/44427#issuecomment-1095923255) an image's "attachment" status isn't perfect for knowing whether it's safe to delete an image from the media library:
- The image could be referenced in a widget or sidebar
- An image will be unattached if it was uploaded to post A, then reused in post B, and then post A is deleted.
- Images that are uploaded directly to the media library and then subsequently used in a post are "correctly" labeled as unattached from a WP point of view.

So I don't think it's safe to run automatically on all sites that were affected by this bug. But depending on the situation it could be a useful tool for Happiness Engineers, see p1656377766709019-slack-C0160HSMDQV for details.

**Comment by CraigG58** [comment]
Thanks! At least that’s some help, and I certainly appreciate your efforts. I assume I need to contact support via normal channels to get your script run on my account?

It’s a shame the first two situations you mentioned aren’t handled automatically and immediately by some sort of back-end server code, since in those scenarios media files are still incorrectly flagged as unattached.

I can confirm you’re 100% correct on the first situation, as my media library contains image files which were recently added to a sidebar that do indeed display as unattached. As for the third scenario, that’s understandable and should be acceptable to live with for most folks.

Anyway, seems a significant oversight in WordPress itself, but I guess it is what it is. The fact remains the media library dialog can’t be trusted if even one file is incorrectly flagged, basically rendering it useless as far as attachment status goes.

Hate to sound like a broken record, but once again that could easily lead to accidental media file deletion by users who aren’t as diligent in monitoring their content as I am, as you’ve also noted.
